File: worldbuilder/geometry/tangent.py
"""
A flat chart of a small piece of a round planet.

This is the file that lets a flat simulation sail on a sphere, and it is not a compromise
made to avoid work - it is what a chart *is*. Every real chart projects a curved surface
onto flat paper; every navigator works plane sailing over short distances and reaches for
great circles only on an ocean passage. A `TangentFrame` is that projection, with an
origin somewhere on the globe and east, north and up defined there.

**The projection is azimuthal equidistant**, chosen for one property: distance and bearing
*from the frame's origin* are exact, at any range, by construction. Nothing is
approximated near the middle of a chart and nothing degrades gracefully - it is simply
right there, and the error is entirely in how two points away from the origin relate to
each other. That is the trade a navigator would recognise, and section 11 of the spec
asks for the error to be measured rather than assumed, which `debug/projection_error.py`
does.

**Nothing outside this package needs to know it exists.** Maritime goes on working in flat
metres and asking how deep the water is; the provider holds the frame and does the
conversion. That is the whole of the integration.
"""

import math
from dataclasses import dataclass

from .sphere import EARTH_RADIUS_M, SpherePoint
from .vectors import DEGENERATE, NORTH_AXIS, POLAR_FALLBACK, Vec3


@dataclass(frozen=True)
class TangentFrame:
    """
    A local flat coordinate system, tangent to the planet at one point.

    Attributes:
        origin (SpherePoint): Where the chart touches the globe; local (0, 0).
        east (Vec3): Unit vector, increasing local x.
        north (Vec3): Unit vector, increasing local y.
        up (Vec3): Unit vector, away from the centre. Equal to the origin's vector.
        radius_m (float): The planet's radius.

    """

    origin: SpherePoint
    east: Vec3
    north: Vec3
    up: Vec3
    radius_m: float

    @classmethod
    def at(cls, origin, radius_m=EARTH_RADIUS_M):
        """
        Build a frame at a point on the globe.

        Args:
            origin (SpherePoint): Where the chart is centred.
            radius_m (float, optional): The planet's radius.

        Returns:
            frame (TangentFrame): A frame with an orthonormal basis, everywhere.

        Notes:
            East is the direction at right angles to both straight up and the planet's
            axis, which is what east means anywhere it means anything.

            **At a pole it means nothing**, and that is not a failure of the maths but a
            fact about poles: every direction from the north pole is south, and no
            direction is east. The cross product goes to zero there and the basis cannot
            be derived, so one is chosen instead - a fixed reference direction, used
            every time. Which direction it is does not matter in the slightest. That it
            is the *same* one on every call is the whole requirement, because a frame
            that reshuffled itself between two calls would move every ship it held.

        """
        up = origin.vector
        sideways = NORTH_AXIS.cross(up)
        if sideways.length() <= DEGENERATE:
            # At a pole, or near enough that the arithmetic has lost its nerve.
            sideways = POLAR_FALLBACK.cross(up)
            if sideways.length() <= DEGENERATE:
                # The fallback was itself parallel to up, which cannot happen for a
                # planet whose axis is z - but a fixed second answer costs one line and
                # removes the only path here that could ever raise.
                sideways = Vec3(0.0, 1.0, 0.0).cross(up)
        east = sideways.normalised()
        north = up.cross(east)
        return cls(origin=origin, east=east, north=north, up=up, radius_m=radius_m)

    @classmethod
    def at_latlon(cls, latitude_deg, longitude_deg, radius_m=EARTH_RADIUS_M):
        """Convenience: a frame centred on a named latitude and longitude."""
        return cls.at(SpherePoint.from_latlon(latitude_deg, longitude_deg), radius_m)

    def local_to_sphere(self, x_m, y_m):
        """
        Where a point on the chart actually is.

        Args:
            x_m (float): Metres east of the frame's origin.
            y_m (float): Metres north of the frame's origin.

        Returns:
            point (SpherePoint): The place on the globe.

        Notes:
            The local distance from the origin is taken as an arc along the surface, not
            as a straight line across the tangent plane - which is what makes the
            projection equidistant, and what stops a thousand-mile chart quietly claiming
            more ocean than the planet has.

        """
        distance = math.hypot(x_m, y_m)
        if distance == 0.0:
            return self.origin

        angle = distance / self.radius_m
        heading = self.east.scaled(x_m / distance) + self.north.scaled(y_m / distance)
        return SpherePoint.from_vector(
            self.up.scaled(math.cos(angle)) + heading.scaled(math.sin(angle))
        )

    def sphere_to_local(self, point):
        """
        Where a place on the globe falls on this chart.

        Args:
            point (SpherePoint): Somewhere on the planet.

        Returns:
            local (tuple): Metres east and north of the frame's origin.

        Notes:
            The exact inverse of `local_to_sphere`, and tested as one. A point directly
            opposite the origin has no direction on this chart at all - every bearing
            reaches it - and returns the origin rather than raising, because a chart of
            half a planet is a misuse the caller should not have to guard against.

        """
        along = self.up.dot(point.vector)
        sideways = point.vector - self.up.scaled(along)
        across = sideways.length()
        if across <= DEGENERATE:
            # The origin itself, or its antipode.
            return 0.0, 0.0

        heading = sideways.scaled(1.0 / across)
        distance = math.atan2(across, along) * self.radius_m
        return heading.dot(self.east) * distance, heading.dot(self.north) * distance

File: crates/worldbuilder-engine/src/tangent.rs
//! A local flat coordinate system, tangent to the planet at one point.
//!
//! Ported from `worldbuilder/geometry/tangent.py`. This is how a curved planet becomes a
//! flat chart in metres and back: maritime works in local metres and never learns the
//! world is a sphere, and the shelf shaper walks fixed distances along geodesics rather
//! than nudging raw coordinates, which would step off the sphere entirely.

use crate::detmath as m;
use crate::sphere::SpherePoint;
use crate::vectors::{Vec3, DEGENERATE, NORTH_AXIS, POLAR_FALLBACK};

#[derive(Debug, Clone, Copy)]
pub struct TangentFrame {
    /// Where the chart touches the globe; local (0, 0).
    pub origin: SpherePoint,
    /// Unit vector, increasing local x.
    pub east: Vec3,
    /// Unit vector, increasing local y.
    pub north: Vec3,
    /// Unit vector, away from the centre. Equal to the origin's vector.
    pub up: Vec3,
    pub radius_m: f64,
}

impl TangentFrame {
    /// East is the direction at right angles to both straight up and the planet's axis.
    ///
    /// At a pole east means nothing — every direction from the north pole is south — and
    /// that is a fact about poles rather than a failure of the maths. The cross product
    /// goes to zero there and the basis cannot be derived, so one is chosen instead.
    /// Which direction it is does not matter in the slightest. That it is the *same* one
    /// on every call is the whole requirement.
    pub fn at(origin: &SpherePoint, radius_m: f64) -> Self {
        let up = origin.vector;
        let mut sideways = NORTH_AXIS.cross(&up);
        if sideways.length() <= DEGENERATE {
            // At a pole, or near enough that the arithmetic has lost its nerve.
            sideways = POLAR_FALLBACK.cross(&up);
            if sideways.length() <= DEGENERATE {
                // The fallback was itself parallel to up, which cannot happen for a
                // planet whose axis is z — but a fixed second answer costs one line and
                // removes the only path here that could ever fail.
                sideways = Vec3::new(0.0, 1.0, 0.0).cross(&up);
            }
        }
        // The Python calls .normalised() and would raise on a zero vector; by this point
        // the fallback chain has guaranteed a non-zero result, so the None case is
        // unreachable. Falling back to POLAR_FALLBACK rather than panicking keeps the
        // function total, and the conformance corpus covers the poles.
        let east = sideways
            .normalised()
            .expect("the fallback chain guarantees a non-zero sideways vector");
        let north = up.cross(&east);
        Self { origin: *origin, east, north, up, radius_m }
    }

    /// Convenience: a frame centred on a named latitude and longitude.
    pub fn at_latlon(latitude_deg: f64, longitude_deg: f64, radius_m: f64) -> Self {
        Self::at(&SpherePoint::from_latlon(latitude_deg, longitude_deg), radius_m)
    }

    /// Where a point on the chart actually is.
    ///
    /// The local distance is taken as an arc along the surface rather than a straight
    /// line across the tangent plane, which is what makes the projection equidistant.
    ///
    /// Written out in components rather than in vector algebra, which is not how the rest
    /// of this crate is written and so needs the excuse: the Python's profiling put this
    /// at a quarter of a chart redraw's cost, it is called six times per terrain sample,
    /// and the tidy version built seven intermediate vectors each time. Same operations,
    /// same order.
    pub fn local_to_sphere(&self, x_m: f64, y_m: f64) -> SpherePoint {
        let distance = m::hypot(x_m, y_m);
        if distance == 0.0 {
            return self.origin;
        }

        let angle = distance / self.radius_m;
        let along = x_m / distance;
        let across = y_m / distance;
        let (east, north, up) = (self.east, self.north, self.up);
        let heading_x = east.x * along + north.x * across;
        let heading_y = east.y * along + north.y * across;
        let heading_z = east.z * along + north.z * across;

        let forward = m::cos(angle);
        let outward = m::sin(angle);
        let x = up.x * forward + heading_x * outward;
        let y = up.y * forward + heading_y * outward;
        let z = up.z * forward + heading_z * outward;

        let scale = 1.0 / m::sqrt(x * x + y * y + z * z);
        SpherePoint { vector: Vec3::new(x * scale, y * scale, z * scale) }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::sphere::{SpherePoint, EARTH_RADIUS_M};
    use crate::vectors::Vec3;

    #[test]
    fn the_basis_is_orthonormal_at_an_ordinary_place() {
        let frame = TangentFrame::at_latlon(51.5, -0.12, EARTH_RADIUS_M);
        assert!((frame.east.length() - 1.0).abs() < 1e-12);
        assert!((frame.north.length() - 1.0).abs() < 1e-12);
        assert!((frame.up.length() - 1.0).abs() < 1e-12);
        assert!(frame.east.dot(&frame.north).abs() < 1e-12);
        assert!(frame.east.dot(&frame.up).abs() < 1e-12);
        assert!(frame.north.dot(&frame.up).abs() < 1e-12);
    }

    #[test]
    fn east_points_east_on_the_equator() {
        // At (0, 0) the up vector is +x, so east must be +y.
        let frame = TangentFrame::at_latlon(0.0, 0.0, EARTH_RADIUS_M);
        assert!((frame.east.y - 1.0).abs() < 1e-12, "east was {:?}", frame.east);
    }

    #[test]
    fn a_pole_still_yields_an_orthonormal_basis() {
        for lat in [90.0, -90.0] {
            let frame = TangentFrame::at_latlon(lat, 0.0, EARTH_RADIUS_M);
            assert!((frame.east.length() - 1.0).abs() < 1e-9, "at {}", lat);
            assert!((frame.north.length() - 1.0).abs() < 1e-9, "at {}", lat);
            assert!(frame.east.dot(&frame.up).abs() < 1e-9, "at {}", lat);
        }
    }

    #[test]
    fn a_pole_yields_the_same_basis_every_time() {
        // The whole requirement. A frame that reshuffled itself between two calls would
        // move every ship it held.
        let a = TangentFrame::at(&SpherePoint { vector: Vec3::new(0.0, 0.0, 1.0) }, EARTH_RADIUS_M);
        let b = TangentFrame::at(&SpherePoint { vector: Vec3::new(0.0, 0.0, 1.0) }, EARTH_RADIUS_M);
        assert_eq!(a.east.x.to_bits(), b.east.x.to_bits());
        assert_eq!(a.east.y.to_bits(), b.east.y.to_bits());
        assert_eq!(a.east.z.to_bits(), b.east.z.to_bits());
    }

    #[test]
    fn up_is_the_origins_own_vector() {
        let origin = SpherePoint::from_latlon(31.0, 7.0);
        let frame = TangentFrame::at(&origin, EARTH_RADIUS_M);
        assert_eq!(frame.up.x.to_bits(), origin.vector.x.to_bits());
        assert_eq!(frame.up.y.to_bits(), origin.vector.y.to_bits());
        assert_eq!(frame.up.z.to_bits(), origin.vector.z.to_bits());
    }

    #[test]
    fn the_origin_maps_to_itself() {
        let frame = TangentFrame::at_latlon(51.5, -0.12, EARTH_RADIUS_M);
        let there = frame.local_to_sphere(0.0, 0.0);
        assert_eq!(there.vector.x.to_bits(), frame.origin.vector.x.to_bits());
        assert_eq!(there.vector.y.to_bits(), frame.origin.vector.y.to_bits());
        assert_eq!(there.vector.z.to_bits(), frame.origin.vector.z.to_bits());
    }

    #[test]
    fn a_step_east_lands_east() {
        let frame = TangentFrame::at_latlon(0.0, 0.0, EARTH_RADIUS_M);
        let (_, lon) = frame.local_to_sphere(100_000.0, 0.0).to_latlon();
        assert!(lon > 0.0, "stepping east gave longitude {}", lon);
    }

    #[test]
    fn the_projection_is_equidistant() {
        // The defining property: local distance equals great-circle distance from the
        // origin. If this drifts, a chart is claiming more ocean than the planet has.
        let frame = TangentFrame::at_latlon(20.0, 30.0, EARTH_RADIUS_M);
        for metres in [1_000.0, 25_000.0, 200_000.0, 1_000_000.0] {
            let there = frame.local_to_sphere(metres, 0.0);
            let measured = frame.origin.distance_to(&there, EARTH_RADIUS_M);
            let error = (measured - metres).abs();
            assert!(error < 1e-6, "at {} m the arc measured {} m", metres, measured);
        }
    }

    #[test]
    fn the_result_is_a_unit_vector() {
        let frame = TangentFrame::at_latlon(-40.0, 170.0, EARTH_RADIUS_M);
        let there = frame.local_to_sphere(300_000.0, -200_000.0);
        assert!((there.vector.length() - 1.0).abs() < 1e-12);
    }
}
